domains/transfer/TransferAgreementAutomation/Worker/Service/Engine/TransferEngineUtility.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnergyOrigin.Domain.ValueObjects;
+using ProjectOriginClients;
+using ProjectOriginClients.Models;
+using TransferAgreementAutomation.Worker.Service.TransactionStatus;
+using RequestStatus = ProjectOriginClients.RequestStatus;
+
+namespace TransferAgreementAutomation.Worker.Service.Engine;
+
+public class TransferEngineUtility
+{
+    public int BatchSize { get; init; } = 1000;
+    public int StatusUpdateIntervalMinutes { get; } = 1;
+
+    private readonly IProjectOriginWalletClient _walletClient;
+    private readonly IRequestStatusRepository _requestStatusRepository;
+
+    public TransferEngineUtility(IProjectOriginWalletClient walletClient, IRequestStatusRepository requestStatusRepository)
+    {
+        _walletClient = walletClient;
+        _requestStatusRepository = requestStatusRepository;
+    }
+
+    public async Task<List<GranularCertificate>> GetCertificates(OrganizationId organizationId)
+    {
+        var hasMoreCertificates = true;
+        var certificates = new List<GranularCertificate>();
+        while (hasMoreCertificates)
+        {
+            var response = await _walletClient.GetGranularCertificates(organizationId.Value, new CancellationToken(), limit: BatchSize,
+                skip: certificates.Count);
+
+            if (response == null)
+            {
+                throw new TransferCertificatesException(
+                    $"Something went wrong when getting certificates from the wallet for {organizationId.Value}. Response is null.");
+            }
+
+            certificates.AddRange(response.Result);
+            if (certificates.Count >= response.Metadata.Total)
+            {
+                hasMoreCertificates = false;
+            }
+        }
+
+        return certificates;
+    }
+
+    public async Task<bool> HasPendingTransactions(OrganizationId organizationId, CancellationToken cancellationToken = default)
+    {
+        var transactionRequests = await _requestStatusRepository.GetByOrganization(organizationId, cancellationToken);
+        await UpdateStatusFromWallet(organizationId, cancellationToken, transactionRequests);
+        return transactionRequests.Any(t => t.Status == Status.Pending);
+    }
+
+    private async Task UpdateStatusFromWallet(OrganizationId organizationId, CancellationToken cancellationToken, IList<TransactionStatus.RequestStatus> transactionRequests)
+    {
+        var transactionsToUpdate = transactionRequests
+            .Where(rs => rs.Status == Status.Pending)
+            .Where(rs => rs.StatusTimestamp < UnixTimestamp.Now().AddMinutes(-StatusUpdateIntervalMinutes))
+            .ToList();
+
+        foreach (var transactionToUpdate in transactionsToUpdate)
+        {
+            var updatedStatus = await _walletClient.GetRequestStatus(organizationId.Value, transactionToUpdate.RequestId, cancellationToken);
```

Comment:
I can see this as a bottleneck sending 100 of requests to get status from wallet. Could be a small DDOS of our selfs? Bulk endpoint would be nice in wallet.

Reply:
Maybe logger error if "createdAt" is bigger than 7 days, so we can see stuff we waited 7 days for to finish? Maybe not an issue, since it's in memory and chances are we deploy more often than that? Just so this list doesn't go wild like the 100.000+ pending in database.

Reply:
The UpdateStatusFromWallet method now deletes any transactions over a day old. But the issue with possible DOS of Wallet is still valid.